src/Nethermind/Nethermind.Evm/Precompiles/Stateful/BeaconBlockRootPrecompile.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.State;
+
+namespace Nethermind.Evm.Precompiles.Stateful;
+public class BeaconBlockRootPrecompile : IPrecompile<BeaconBlockRootPrecompile>
+{
+    public static Address Address { get; } = Address.FromNumber(0x0B);
+    public static UInt256 HISTORICAL_ROOTS_LENGTH = 98304;
+
+    public static BeaconBlockRootPrecompile Instance => new BeaconBlockRootPrecompile();
+
+    private byte[] SloadFromStorage(IWorldState state, in UInt256 index)
+    {
+        StorageCell storageCell = new(Address, index);
+        return state.Get(storageCell);
+    }
+    public void SetupBeaconBlockRootPrecompileState(bool isEip4788Enabled, IWorldState stateProvider, Keccak parentBeaconBlockRoot, UInt256 timestamp)
+    {
+        if (!isEip4788Enabled) return;
+        stateProvider.CreateAccountIfNotExists(BeaconBlockRootPrecompile.Address, 1);
+
+        if (parentBeaconBlockRoot is null) return;
+
+        UInt256.Mod(timestamp, HISTORICAL_ROOTS_LENGTH, out UInt256 timestampReduced);
+        UInt256 rootIndex = timestampReduced + HISTORICAL_ROOTS_LENGTH;
+
+        StorageCell tsStorageCell = new(Address, timestampReduced);
+        StorageCell brStorageCell = new(Address, rootIndex);
+
+        stateProvider.Set(tsStorageCell, timestamp.ToBigEndian());
+        stateProvider.Set(brStorageCell, parentBeaconBlockRoot.Bytes.ToArray());
```

Comment:
Idea - maybe we could change `Set` to accept `ReadOnlyMemory<byte>` instead of `byte[]` and reduce allocations?
@benaadams @asdacap